example-app/src/main/java/com/adyen/checkout/example/ui/giftcard/GiftCardViewModel.kt

```diff
@@ -0,0 +1,277 @@
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.filterNotNull
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import org.json.JSONObject
+import javax.inject.Inject
+
+@HiltViewModel
+@Suppress("TooManyFunctions")
+internal class GiftCardViewModel @Inject constructor(
+    private val savedStateHandle: SavedStateHandle,
+    private val paymentsRepository: PaymentsRepository,
+    private val keyValueStorage: KeyValueStorage,
+) : ViewModel(), GiftCardComponentCallback {
+
+    private val _giftCardComponentDataFlow = MutableStateFlow<GiftCardComponentData?>(null)
+    val giftCardComponentDataFlow: Flow<GiftCardComponentData> = _giftCardComponentDataFlow.filterNotNull()
+
+    private val _giftCardViewStateFlow = MutableStateFlow<GiftCardViewState>(GiftCardViewState.Loading)
+    internal val giftCardViewStateFlow: Flow<GiftCardViewState> = _giftCardViewStateFlow
+
+    private val _events = MutableSharedFlow<GiftCardEvent>()
+    internal val events: Flow<GiftCardEvent> = _events
+
+    private var order: OrderRequest? = null
+
+    init {
+        viewModelScope.launch { fetchPaymentMethods() }
+    }
+
+    private suspend fun fetchPaymentMethods() = withContext(Dispatchers.IO) {
+        val paymentMethodResponse = paymentsRepository.getPaymentMethods(
+            getPaymentMethodRequest(
+                merchantAccount = keyValueStorage.getMerchantAccount(),
+                shopperReference = keyValueStorage.getShopperReference(),
+                amount = keyValueStorage.getAmount(),
+                countryCode = keyValueStorage.getCountry(),
+                shopperLocale = keyValueStorage.getShopperLocale(),
+                splitCardFundingSources = keyValueStorage.isSplitCardFundingSources()
+            )
+        )
+
+        val giftCardPaymentMethod = paymentMethodResponse
+            ?.paymentMethods
+            ?.firstOrNull { GiftCardComponent.PROVIDER.isPaymentMethodSupported(it) }
+
+        if (giftCardPaymentMethod == null) {
+            _giftCardViewStateFlow.emit(GiftCardViewState.Error)
+        } else {
+            _giftCardComponentDataFlow.emit(
+                GiftCardComponentData(
+                    paymentMethod = giftCardPaymentMethod,
+                    callback = this@GiftCardViewModel,
+                )
+            )
+            _giftCardViewStateFlow.emit(GiftCardViewState.ShowComponent)
+        }
+    }
+
+    override fun onSubmit(state: GiftCardComponentState) {
+        makePayment(state.data)
+    }
+
+    override fun onAdditionalDetails(actionComponentData: ActionComponentData) {
+        sendPaymentDetails(actionComponentData)
+    }
+
+    override fun onError(componentError: ComponentError) {
+        onComponentError(componentError)
+    }
+
+    // no ops
+    override fun onStateChanged(state: GiftCardComponentState) = Unit
```

Comment:
This is optional, right? Are you overriding it just to show it exists?

Reply:
Indeed.